### ISSUE TITLE
[Bug]: Segfault if Reconciler hooks registered, but either Pre or Post hooks not implemented

### ISSUE THREAD

**Issue by ringerc** [cloudnative-pg/cnpg-i-machinery#32]
### Is there an existing issue already for this bug?

- [X] I have searched for an existing issue, and could not find anything. I believe this is a new bug.

### I have read the troubleshooting guide

- [X] I have read the troubleshooting guide and I think this is a new bug.

### What happened?

I wrote a reconciler hook that registered a Pre hook callback.

It ran, then segfaulted after returning.

The cause is that the Post hook must also be implemented. Otherwise the plugin helper tries to call a null function pointer.

Plugin helper should check that both Pre and Post hooks are non-null and error at registration time. Or it should skip unimplemented hooks.

This was made confusing by the panic handler's swallowing of the stack trace, so all I got was the `"error": "panic: runtime error: invalid memory address or nil pointer dereference"` but no details. See https://github.com/cloudnative-pg/cnpg-i-machinery/issues/33

### Relevant log output

```shell
{"level":"info","ts":1719969643.970347,"caller":"pluginhelper/server.go:195","msg":"Panic occurred","error":"runtime error: invalid memory address or nil pointer dereference"}
```

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by armru** [comment]
Hey @ringerc, what would be the best solution for you?

**Comment by armru** [comment]
Hello again, did also try to embbed UnimplementedReconcilerHooksServer?

**Comment by ringerc** [comment]
I did not - I used the example plugin as a base. I didn't see this anywhere in the docs and examples at the time I was working on it.

I expected that the helper would check if the routine was implemented and gracefully, silently skip it if not.

I now see that the other implementations do have an unimplemented helper base:

```
type Implementation struct {
	lifecycle.UnimplementedOperatorLifecycleServer
}
```

whereas mine uses the server type:

```
type ReconcileHandler struct {
	reconciler.ReconcilerHooksServer
}
```

It would be helpful if the server types had prominent references to the Unimplemented bases in their godoc. A comment in the hello world to mention why it's used wouldn't hurt either.

Where does that come from anyway? https://github.com/search?q=org%3Acloudnative-pg+UnimplementedOperatorLifecycleServer&type=code - I don't see it defined anywhere.

**Comment by armru** [comment]
Yes it is autogenerated and yes we should have a better reference to it :)

**Comment by ringerc** [comment]
Yes, that was correct. Using `lifecycle.UnimplementedOperatorLifecycleServer` as the base solved the issue, it's just a docs issue.
